src/ansi.rs

```diff
@@ -0,0 +1,2089 @@
+            None => self.advance_sync_dcs_start(),
+        }
+    }
+
+    /// Find the start of sync DCS sequences.
+    fn advance_sync_dcs_start(&mut self) {
+        // Get the last few bytes for comparison.
+        let len = self.state.sync_state.buffer.len();
+        let offset = len.saturating_sub(SYNC_ESCAPE_START_LEN);
+        let end = &self.state.sync_state.buffer[offset..];
+
+        // Check for extension/termination of the synchronized update.
+        if end == SYNC_START_ESCAPE_START {
+            self.state.sync_state.pending_dcs = Some(Dcs::SyncStart);
+        } else if end == SYNC_END_ESCAPE_START || len >= SYNC_BUFFER_SIZE - 1 {
+            self.state.sync_state.pending_dcs = Some(Dcs::SyncEnd);
+        }
+    }
+
+    /// Parse the DCS termination sequence for synchronized updates.
+    fn advance_sync_dcs_end<H>(&mut self, handler: &mut H, byte: u8)
+    where
+        H: Handler,
+    {
+        match byte {
+            // Ignore DCS passthrough characters.
+            0x00..=0x17 | 0x19 | 0x1c..=0x7f | 0xa0..=0xff => (),
+            // Cancel the DCS sequence.
+            0x18 | 0x1a | 0x80..=0x9f => self.state.sync_state.pending_dcs = None,
+            // Dispatch on ESC.
+            0x1b => match self.state.sync_state.pending_dcs.take() {
+                Some(Dcs::SyncStart) => {
+                    self.state.sync_state.timeout.set_timeout(SYNC_UPDATE_TIMEOUT);
+                },
+                Some(Dcs::SyncEnd) => self.stop_sync(handler),
+                None => (),
+            },
+        }
+    }
+}
+
+/// Helper type that implements `crate::Perform`.
+///
+/// Processor creates a Performer when running advance and passes the Performer
+/// to `crate::Parser`.
+struct Performer<'a, H: Handler, T: Timeout> {
+    state: &'a mut ProcessorState<T>,
+    handler: &'a mut H,
+}
+
+impl<'a, H: Handler + 'a, T: Timeout> Performer<'a, H, T> {
+    /// Create a performer.
+    #[inline]
+    pub fn new<'b>(state: &'b mut ProcessorState<T>, handler: &'b mut H) -> Performer<'b, H, T> {
+        Performer { state, handler }
+    }
+}
+
+#[cfg(not(feature = "no_std"))]
+#[derive(Default)]
+pub struct StdSyncHandler {
+    timeout: Option<Instant>,
+}
+
+#[cfg(not(feature = "no_std"))]
+impl StdSyncHandler {
+    /// Synchronized update expiration time.
+    #[inline]
+    pub fn sync_timeout(&self) -> Option<Instant> {
+        self.timeout
+    }
+}
+
+#[cfg(not(feature = "no_std"))]
+impl Timeout for StdSyncHandler {
+    #[inline]
+    fn set_timeout(&mut self, duration: Duration) {
+        self.timeout = Some(Instant::now() + duration);
+    }
+
+    #[inline]
+    fn clear_timeout(&mut self) {
+        self.timeout = None;
+    }
+
+    #[inline]
+    fn pending_timeout(&self) -> bool {
+        self.timeout.is_some()
+    }
+}
+
+/// Provides an interface for common timer functionality (i.e.
+/// creating timeouts and checking if it has been expired).
```

Comment:
```suggestion
/// Interface for creating timeouts and checking their expiry.
```

The whole parenthesis situation probably doesn't generate the best docs and reads a bit awkward. I think this one simple line should say the same thing but look a bit cleaner.